### ISSUE TITLE
Prompt caching support & Potential improvements

### ISSUE THREAD

**Issue by p4-k4** [pasky/claude.vim#16]
First of all, thanks for bringing this to life!

Unsure exactly if prompt caching is implemented under the hood however, I'd hope so! There's some huge savings $$ to be made.

[https://docs.anthropic.com/en/docs/build-with-claude/prompt-caching](https://docs.anthropic.com/en/docs/build-with-claude/prompt-caching)

It would also be nice to have feedback on cost, just as [Cline](https://github.com/cline/cline) does. It summarises cost based on a "Task" and also granually on each prompt and response for that task.

Cline has functionality to detect `// Rest of code remains the same` and warn you, then you can manually merge these parts.

The ability to be able to `@tag` files/folders would be great too to have finer control over what's included in a prompt.

Cheers

**Comment by pasky** [comment]
All nice ideas! I have been already thinking about prompt-caching in the past, it would be totally reasonable. It would also let us rely more on tool use for suggesting edits, rather than the markdown blocks which are actually pretty fragile (especially the associated movements).

I'm also unlikely right now to implement this myself, but I'll be happy to review contributions.

There are cost estimates printed as messages currently - there might be some better UX, of course.

@tagging may make sense, though in the way I use vim personally the context aligns pretty well. (If anything, it would be nice to have support for smart languageserver-driven context makeup, perhaps using https://github.com/codestoryai/sidecar or something similar.) But a feature on my wishlist forever is to be able to switch between a "complex" and "simple" prompt - I'd use claude.vim more if it didn't launch into lengthy problem-solving deliberation for every simple question I ask it. :)

**Comment by p4-k4** [comment]
Yeah you're aware of prompt caching in the anthropic API call right? (it's a relatively [new thing](https://www.anthropic.com/news/prompt-caching) and in beta I think) but I saved a lot of both time and money with it, I wouldn't run sonnet without it otherwise too expensive.

I almost (very reluctantly) switched to vscode just because of Cline as vim lacks these kind of plugins, except that `claude.vim` seems to be closest resembling. I'm currently using vscode/cline for time consuming tasks, then switching back to nvim for manual writing although obviously not preferred.

Cline has a very small and simple front-facing feature set and it functions extremely well without being over-complicated, but its core logic is quite powerful - If you get a chance, have a look, might spark some ideas.

Pretty sure there would be some pre-defined instructions/prompts, as it also pulls things such as diagnostics, terminal output etc from vscode, it has look-ahead on mistakes and fixes them (extremely well) etc.

My experience with @tagging filesfolders in Cline is that we can selectively reduce tokens that go out and also restrict Cline to only working on those particular files for that given task. Otherwise it just reads files as it gathers context (adding to the prompt cache).

I'm far from lua adept myself unfortunately but hopefully someone else picks this up or if I get desperate I'll start digging.

**Comment by pasky** [comment]
It's not hard to get going if you use claude.vim to extend claude.vim, even if you don't know any vimscript or Python! (Though don't expect Claude to do everything on its own either.)

I don't have a lot of time to develop new Claude.vim features at the time being, I'm usually improving it only in short burts when I use it intensely, which isn't really "daily" right now due to my work mix.

**Comment by danhorner** [comment]
I too  have been poking a little at prompt-caching.  I want to (1) load documentation and associated code and then (2) cache conversational turns.  I am trying to modify claude.vim so that it caches the system prompt and any shared bufffers. 

But I trip over the case when claude suggests edits to open buffers and invalidates the cache. 

How about this approach? 
1. Set a cache point for the system prompt and tool definitions
2. Move the shared buffers from the system prompt into a `{ content: [] }` block in the first user message.
3. Sort the buffers so that read-only buffers appear first in the content.  Apply a `cache-control` header to the last read-only buffer. 
4. Also apply a `cache-control` header to the last user message in the conversation. 

The first cache-control block should preserve large source files that don't change, even if claude suggests edits to other files. The second cache-control block should preserve all files plus any conversation when claude doesn't suggest edits.
